pallets/credit/src/lib.rs

```diff
@@ -728,17 +837,64 @@ pub mod pallet {
             if new_id.is_none() {
                 return false;
             }
+            let new_id = new_id.unwrap();
 
-            old_data.campaign_id = new_id.unwrap();
-            old_data.reward_eras += 180;
+            if old_data.campaign_id == new_id {
+                old_data.reward_eras += 180;
+            } else {
+                old_data.campaign_id = new_id;
+                old_data.reward_eras = DEFAULT_REWARD_ERAS;
+            }
 
-            UserCredit::<T>::insert(who, old_data.clone());
+            UserCredit::<T>::insert(who, old_data);
             Self::update_credit_history(who, expire_era);
             true
         }
+
+        fn do_add_credit(account_id: T::AccountId, credit_data: CreditData) -> DispatchResult {
+            if UserCredit::<T>::contains_key(&account_id) {
+                UserCredit::<T>::mutate(&account_id, |d| match d {
+                    Some(data) => *data = credit_data.clone(),
+                    _ => (),
+                });
+                if !Self::user_credit_history(&account_id).is_empty() {
+                    Self::update_credit_history(&account_id, Self::get_current_era());
+                }
+            } else {
+                UserCredit::<T>::insert(&account_id, credit_data.clone());
+            }
+            Self::deposit_event(Event::CreditUpdateSuccess(account_id, credit_data.credit));
+            Ok(())
+        }
     }
 
     impl<T: Config> CreditInterface<T::AccountId, BalanceOf<T>> for Pallet<T> {
+        fn get_credit_balance() -> Vec<BalanceOf<T>> {
+            Self::credit_balances()
+        }
+
+        fn get_credit_gap(dst_lv: u8, cur_lv: u8) -> u64 {
+            CreditLevel::credit_level_gap(dst_lv.into(), cur_lv.into())
+        }
+
+        fn add_or_update_credit(account_id: T::AccountId, credit_score: u64) -> DispatchResult {
```

Comment:
No `Err` was returned, consider not return a `Result`

Reply:
good